File: Kattis/TernarianWeights.hs
import           Control.Applicative   ((<|>))
import           Control.Arrow         ((***), (>>>))
import qualified Data.ByteString.Char8 as C
import           Data.Function         ((&))
import           Data.List             (intercalate)
import           Data.Maybe            (fromJust)
import           Text.Printf           (printf)

main :: IO ()
main = C.getContents >>= (
            C.lines
        >>> drop 1
        >>> map (readInt >>> solve)
        >>> intercalate "\n\n"
        >>> putStrLn
    )

solve :: Int -> String
solve w = distribute ([],[]) w threes
    & fromJust
    & (map show >>> unwords) *** (map show >>> unwords)
    & uncurry (printf "left pan: %s\nright pan: %s")
  where
    distribute :: ([Int],[Int]) -> Int -> [Int] -> Maybe ([Int],[Int])
    distribute (ls,rs) 0 _      = Just (reverse ls, reverse rs)
    distribute _ _ []           = Nothing
    distribute (ls,rs) c (x:xs) | c < 0     = distribute (ls,rs) c xs <|> distribute (x:ls,rs) (c+x) xs
                                | otherwise = distribute (ls,rs) c xs <|> distribute (ls,x:rs) (c-x) xs

threes :: [Int]
threes = map (3^) [20,19..0]

readInt :: C.ByteString -> Int
readInt = C.readInt >>> fromJust >>> fst

File: Kattis/ternarianweights.py
THREES = [3**i for i in range(20)]

def make(w,i):
    if w == 0: return ('', '')
    if i < 0: return None

    skip = make(w,i-1)
    if skip is not None:
        return skip
    if w < 0:
        left = make(w+THREES[i],i-1)
        if left is not None:
            l,r = left
            return (f'{THREES[i]} {l}', r)
    else:
        right = make(w-THREES[i],i-1)
        if right is not None:
            l,r = right
            return (l, f'{THREES[i]} {r}')

    return None

for c in range(int(input())):
    x = int(input())
    ret = make(x, 20)
    if c != 0:
        print()
    print(f'left pan: {ret[0]}\nright pan: {ret[1]}')
